docs/ocis/adr/0029-grpc-in-kubernetes.md

```diff
@@ -0,0 +1,98 @@
+
+Unfortunately, reva does not use go micro grpc clients. We implemented our own selector mechanism for the reva pool and always select the next client. This gives us an ip that is used by the upstream grpc-go client to dial the connection. No retries are configured, yet.
+
+What makes this worse is that we cannot use the native retry mechanism of grpc-go, because we already looked up an ip and the pod might already have been killed and we would just retry sending requests to the same ip.
+
+To top it all off, go-micro V5 has changed the license to BSL, which raises concerns on how long we can safely rely on it.
+
+We need to decide how we want to load balance and retry grpc connections in kubernetes.
+
+## Decision Drivers
+
+* oCIS should scale in Kubernetes without losing requests
+* The code should be maintainable
+* Connections should work on localhost as well as in kubernetes
+
+## Considered Options
+* go-micro clients
+* Proxy load balancing
+* Thick client-side load balancing
+* Lookaside Load Balancing
+
+## Decision Outcome
+### Positive Consequences:
+### Negative consequences:
+
+## Pros and Cons of the Options
+### go-micro clients
+* good, because we can use the go micro client retry mechanism
+* good, because we keep using interfaces that allow us to change the implementatien to test nats-js-kv vs kubernetes or whatever
+* bad, pod readiness in kubernetes is basically ignored
+* bad, because go-micro v5 changed the licence to BSL
+* bad, because we need to change every line of code in reva that makes a grpc call
+
+### [Proxy load balancing](https://grpc.io/blog/grpc-load-balancing/#proxy-load-balancer-options)
+We could use a L7 proxy like envoy to do the gRPC load balancing.  kubernetes would have to return all pod ips in dns responses by setting `clusterIP: None` to use headless services. And clients would have to use the envoy proxy address
+* good, because clients are simple
+* bad, because proxy adds an extra hop
+* bad, because we would not be using the go-micro service registry
+
+### [Thick client-side load balancing](https://grpc.io/blog/grpc-load-balancing/#thick-client)
+For this we would have to replace the go micro service registry and rely on dns as a service registry. service names would have to be configured with a schema, eg. `dns:///localhost:9142`, `dns:///gateway.ocis.svc.cluster.local:9142` or `unix:/var/run/gateway.socket` and kubernetes would have to return all pod ips in dns responses by setting `clusterIP: None` to use headless services.
+* good, because we can use the grpc-go native retry mechanism
+* good, because pod readyness is respected
+* good, because we get rid of the complexity of a service registry - which means revisiting [ADR0006 Service Discovery](https://owncloud.dev/ocis/adr/0006-service-discovery/)
+* bad, because we would lose the service registry
```

Comment:
I put it as good and bad:
```
* good, because we get rid of the complexity of a service registry - which means revisiting [ADR0006 Service Discovery](https://owncloud.dev/ocis/adr/0006-service-discovery/)
* bad, because we would lose the service registry
```